app/vminsert/netstorage/insert_ctx.go

```diff
@@ -122,15 +122,12 @@ func (ctx *InsertCtx) WriteDataPoint(at *auth.Token, labels []prompb.Label, time
 func (ctx *InsertCtx) WriteDataPointExt(at *auth.Token, storageNodeIdx int, metricNameRaw []byte, timestamp int64, value float64) error {
 	br := &ctx.bufRowss[storageNodeIdx]
 	sn := storageNodes[storageNodeIdx]
-	bufNew := storage.MarshalMetricRow(br.buf, metricNameRaw, timestamp, value)
-	if len(bufNew) >= maxBufSizePerStorageNode {
+	br.buf = storage.MarshalMetricRow(br.buf, metricNameRaw, timestamp, value)
+	if len(br.buf) >= maxBufSizePerStorageNode {
 		// Send buf to storageNode, since it is too big.
 		if err := br.pushTo(sn); err != nil {
 			return err
 		}
-		br.buf = storage.MarshalMetricRow(bufNew[:0], metricNameRaw, timestamp, value)
```

Comment:
This superfluous data copy occurs only when `len(bufNew) < maxBufSizePerStorageNode`, e.g. when the per-vmstorage buffer becomes full. Usually the buffer size is quite big (currently it is limited to 30MB according to `consts.MaxInsertPacketSizeForVMInsert`, while the size of `metricNameRaw` is usually less than 1KB. This means that this superfluous data copy is very rare in practice, so there is no need in optimizing it.